CP-405 Further refactor for invoicing system

ecino opened the pull request:
- [FIX] More granularity when updating a sponsorship. Prevents a change in one sponsorship to affect all other invoice lines in the same invoice.
- Move code to better isolate gift functionality inside sponsorship_compassion module instead of recurring_contract.

ecino committed:
CP-405 Further refactor for invoicing system

- [FIX] More granularity when updating a sponsorship.
  Prevents a change in one sponsorship to affect all
  other invoice lines in the same invoice.
- Move code to better isolate gift functionality inside
  sponsorship_compassion module instead of recurring_contract.

simgonzalez commented on res_partner_compassion/models/res_partner.py:
```diff
@@ -26,12 +26,6 @@ def _updt_invoices_rp(self, vals):
-                            payment_term_id=self.property_payment_term_id.id
-                        )
-                    )
+            data_invs = invoices._build_invoice_data(payment_term_id=self.property_payment_term_id.id)
```
```suggestion
            data_invs = invoices._build_invoices_data(payment_term_id=self.property_payment_term_id.id)
```

simgonzalez commented on sponsorship_compassion/models/account_move.py:
```diff
@@ -63,3 +63,11 @@ def _compute_invoice_category(self):
         self._compute_invoice_category()
+
+    def _build_invoice_lines_from_contracts(self, modified_contracts):
+        if self.invoice_category == "sponsorship":
+            return super()._build_invoice_lines_from_contracts(modified_contracts)
+        else:
+            # Handle a gift or fund change
+            return super(AccountInvoice, self.with_context(
+                open_invoices_exclude_sponsorship=True))._build_invoice_lines_from_contracts(modified_contracts)
```
```suggestion
        if self.invoice_category == "sponsorship":
            super = super()
        else:
            # Handle a gift or fund change
            super = super(AccountInvoice, self.with_context(
                open_invoices_exclude_sponsorship=True))
        return super._build_invoice_lines_from_contracts(modified_contracts) 
```

simgonzalez commented on sponsorship_compassion/models/account_move_line.py:
```diff
@@ -0,0 +1,42 @@
+            for invoice_line in self:
+                data_dict = {}
+                # Process specific cases for gift
+                if invoice_line.product_id.default_code == PRODUCT_GIFT_CHRISTMAS:
+                    data_dict["price_unit"] = modified_contract.christmas_invoice
+                elif invoice_line.product_id.default_code == GIFT_PRODUCTS_REF[0]:
+                    data_dict["price_unit"] = modified_contract.birthday_invoice
+                else:
+                    raise UserError(_("Unexpected error while updating contract invoices. Please contact admin."))
```
```suggestion
                if invoice_line.product_id.default_code == PRODUCT_GIFT_CHRISTMAS:
                    gift_type = 'christmas'
                elif invoice_line.product_id.default_code == GIFT_PRODUCTS_REF[0]:
                    gift_type = 'birthday'
                else:
                    raise UserError(_("Unexpected error while updating contract invoices. Please contact admin."))
                # Assign the price depending on the gift type
                data_dict["price_unit"] = eval('modified_contract.' + gift_type + '_invoice')
```

ecino replied:
```suggestion
                if invoice_line.product_id.default_code == PRODUCT_GIFT_CHRISTMAS:
                    gift_type = 'christmas_invoice'
                elif invoice_line.product_id.default_code == GIFT_PRODUCTS_REF[0]:
                    gift_type = 'birthday_invoice'
                else:
                    raise UserError(_("Unexpected error while updating contract invoices. Please contact admin."))
                # Assign the price depending on the gift type
                data_dict["price_unit"] = getattr(modified_contract, gift_type)
```

simgonzalez commented on sponsorship_compassion/models/contracts.py:
```diff
@@ -1109,3 +1134,15 @@ def cancel_old_invoices(self):
+        if "birthday_invoice" in vals or "christmas_invoice" in vals:
+            contracts = self.with_context(open_invoices_exclude_sponsorship=True)
+            gifts = contracts.mapped("open_invoice_ids")
+            data_invs = gifts._build_invoice_data(contracts=contracts)
```
```suggestion
            data_invs = gifts._build_invoices_data(contracts=contracts)
```

ecino committed:
Apply suggestions from code review

Co-authored-by: Simon Gonzalez <45272875+simgonzalez@users.noreply.github.com>

ecino committed:
CP-405 Code review and further bug fix

- If a gift was present in the future when creating
  a new sponsorship, the invoice generation found
  the gift invoice line and updated it by adding
  the sponsorship line inside the gift invoice.
  The FIX exclude now gifts from open invoices consideration when generating the invoices of a contract.
